src/compiler/checker.ts

```diff
@@ -40559,6 +40560,29 @@ namespace ts {
             }
         }
 
+        function checkGrammarJsxName(node: JsxTagNameExpression) {
+            if (isPropertyAccessExpression(node)) {
+                let propName: JsxTagNameExpression = node;
+                do {
+                    const check = checkGrammarJsxNestedIdentifier(propName.name);
+                    if (check) {
+                        return check;
+                    }
+                    propName = propName.expression;
+                } while (isPropertyAccessExpression(propName));
+                const check = checkGrammarJsxNestedIdentifier(propName);
+                if (check) {
+                    return check;
+                }
+            }
+
+            function checkGrammarJsxNestedIdentifier(name: MemberName | ThisExpression) {

```

Comment:
This doesn't have a strong reason to be nested, I think.

Reply:
Mostly just to avoid pollution in the containing scope, since it's only needed here. Since it doesn't close over anything, it should be just as effective as a function in the surrounding scope.